src/Network/OWSSignalService.m

```diff
@@ -28,24 +52,105 @@ - (instancetype)init
+
+    OWSSingletonAssert();
+
     return self;
 }
 
-- (BOOL)isCensored
+- (void)observeNotifications
 {
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(registrationStateDidChange:)
+                                                 name:kNSNotificationName_RegistrationStateDidChange
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(localNumberDidChange:)
+                                                 name:kNSNotificationName_LocalNumberDidChange
+                                               object:nil];
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)updateHasCensoredPhoneNumber
+{
+    OWSAssert([NSThread isMainThread]);
+
     NSString *localNumber = [TSAccountManager localNumber];
 
     if (localNumber) {
-        return [self.censorshipConfiguration isCensoredPhoneNumber:localNumber];
+        self.hasCensoredPhoneNumber = [self.censorshipConfiguration isCensoredPhoneNumber:localNumber];
     } else {
-        DDLogError(@"no known phone number to check for censorship.");
-        return NO;
+        DDLogError(@"%@ no known phone number to check for censorship.", self.tag);
+        self.hasCensoredPhoneNumber = NO;
+    }
+
+    [self updateIsCensorshipCircumventionActive];
+}
+
+- (BOOL)isCensorshipCircumventionManuallyActivated
+{
+    OWSAssert([NSThread isMainThread]);
+
+    return [[NSUserDefaults.standardUserDefaults
+        objectForKey:kNSUserDefaults_isCensorshipCircumventionManuallyActivated] boolValue];
+}
+
+- (void)setIsCensorshipCircumventionManuallyActivated:(BOOL)value
+{
+    OWSAssert([NSThread isMainThread]);
+
+    [NSUserDefaults.standardUserDefaults setObject:@(value)
+                                            forKey:kNSUserDefaults_isCensorshipCircumventionManuallyActivated];
+    [NSUserDefaults.standardUserDefaults synchronize];
+
+    [self updateIsCensorshipCircumventionActive];
+}
+
+- (void)updateIsCensorshipCircumventionActive
+{
+    OWSAssert([NSThread isMainThread]);
+
+    self.isCensorshipCircumventionActive
+        = (self.isCensorshipCircumventionManuallyActivated || self.hasCensoredPhoneNumber);
+}
+
+- (void)setIsCensorshipCircumventionActive:(BOOL)isCensorshipCircumventionActive
+{
+    OWSAssert([NSThread isMainThread]);
+
+    @synchronized(self)
+    {
+        if (_isCensorshipCircumventionActive == isCensorshipCircumventionActive) {
+            return;
+        }
+
+        _isCensorshipCircumventionActive = isCensorshipCircumventionActive;
+    }
+
+    [[NSNotificationCenter defaultCenter]
+        postNotificationName:kNSNotificationName_IsCensorshipCircumventionActiveDidChange
+                      object:nil
+                    userInfo:nil];
+}
+
+- (BOOL)isCensorshipCircumventionActive
+{
```

Comment:
Access to the property is `@synchronized` because this might be accessed on any thread.